File: ipad-app/BridgeClient.swift
// iPad 앱과 백엔드 사이의 HTTP 통신을 모아 둠
import Foundation

// MARK: - 서버 응답 모델
enum TaskStatus: String, Decodable {
    case queued, interpreting
    case awaitingConfirmation = "awaiting_confirmation"
    case resolvingSession = "resolving_session"
    case runningAgent = "running_agent"
    case testing, completed, failed, cancelled
    var isActive: Bool {
        switch self {
        case .completed, .failed, .cancelled: return false
        default: return true
        }
    }
}

struct TaskCreated: Decodable {
    let taskId: String
    let status: TaskStatus
    let conversationId: String?
}

struct AgentOption: Decodable, Identifiable {
    let value: String
    let label: String
    var id: String { value }
}

struct AgentView: Decodable, Identifiable {
    let agentId: String
    let displayName: String
    let usable: Bool
    let note: String
    let models: [AgentOption]
    var id: String { agentId }
}

struct ConversationView: Decodable, Identifiable {
    let conversationId: String
    let projectId: String
    let title: String
    let createdAt: Date
    let updatedAt: Date
    let agentSessions: [String: String]
    var id: String { conversationId }
}

struct ConversationDetail: Decodable {
    let conversation: ConversationView
    let tasks: [TaskView]
}

struct ProjectView: Decodable, Identifiable {
    enum Status: String, Decodable { case idle, busy, unavailable }

    let projectId: String
    let displayName: String
    let status: Status
    let activeTaskId: String?
    let reason: String?
    let agent: String
    let previewAvailable: Bool

    var id: String { projectId }
}

struct ChangedFile: Decodable { let path: String; let summary: String }
struct TestResult: Decodable { let command: String; let status: String; let summary: String }
struct QuestionOption: Decodable, Identifiable {
    let optionId: String
    let label: String
    var id: String { optionId }
}
struct OverlayTarget: Decodable {
    let shape: String
    let x: Double
    let y: Double
    let width: Double
    let height: Double
    let label: String
}
struct Question: Decodable, Identifiable {
    let questionId: String
    let text: String
    let options: [QuestionOption]
    let overlay: OverlayTarget?
    var id: String { questionId }
}

struct PreviewView: Decodable {
    let projectId: String
    let url: URL
    let port: Int
}

struct HealthView: Decodable {
    let status: String
    let projects: Int
}

struct TailscaleDeviceView: Decodable, Identifiable {
    let name: String
    let dnsName: String
    let online: Bool
    let isSelf: Bool

    var id: String { dnsName }
    var bridgeURL: URL? { URL(string: "http://\(dnsName):8787") }
}

struct TaskView: Decodable {
    let taskId: String
    let projectId: String
    let conversationId: String?
    let status: TaskStatus
    let sessionId: String?
    let agentId: String?
    let agentModel: String?
    let userMessage: String
    let createdAt: Date
    let updatedAt: Date
    let summary: String?
    let agentReply: String?
    let changedFiles: [ChangedFile]
    let testResults: [TestResult]
    let questions: [Question]
    let warnings: [String]
    let error: String?
}

// MARK: - 오류

struct BridgeError: LocalizedError {
    let message: String
    let statusCode: Int?

    var errorDescription: String? { message }

    static func network(_ error: Error) -> BridgeError {
        let message: String
        switch (error as? URLError)?.code {
        case .some(.cannotConnectToHost), .some(.cannotFindHost):
            message = "iMac에 연결할 수 없습니다. 브리지 서버가 켜져 있는지 확인해 주세요."
        case .some(.notConnectedToInternet), .some(.networkConnectionLost):
            message = "네트워크 연결이 끊겼습니다. 잠시 후 다시 시도해 주세요."
        case .some(.timedOut):
            message = "iMac이 응답하지 않습니다."
        case .some(.appTransportSecurityRequiresSecureConnection):
            message = "평문 HTTP 연결이 차단되었습니다. Info.plist의 ATS 설정을 확인해 주세요."
        default:
            message = error.localizedDescription
        }
        return BridgeError(message: message, statusCode: nil)
    }
}

private struct ErrorDetail: Decodable { let detail: String? }

// MARK: - 클라이언트

/// PC의 Bridge와 통신
struct BridgeClient {
    let baseURL: URL
    var session: URLSession = .shared

    private static let prefix = "api/v1"

    func health() async throws -> HealthView {
        try await send(request(path: "health"), as: HealthView.self)
    }

    func tailscaleDevices() async throws -> [TailscaleDeviceView] {
        struct Response: Decodable { let devices: [TailscaleDeviceView] }
        return try await send(
            request(path: "tailscale/devices"),
            as: Response.self
        ).devices
    }

    /// 사용자가 연결 확인에 쓴 `/api/v1/health` 주소를 그대로 붙여 넣더라도 API 경로가 중복되지 않게 PC 서버의 루트 URL로 되돌림
    private var serverRootURL: URL {
        guard var components = URLComponents(
            url: baseURL, resolvingAgainstBaseURL: false
        ) else { return baseURL }

        var segments = components.path.split(separator: "/").map(String.init)
        if let apiIndex = segments.indices.first(where: { index in
            segments[index].lowercased() == "api"
                && segments.indices.contains(index + 1)
                && segments[index + 1].lowercased() == "v1"
        }) {
            segments = Array(segments[..<apiIndex])
        }
        components.path = segments.isEmpty ? "" : "/" + segments.joined(separator: "/")
        components.query = nil
        components.fragment = nil
        return components.url ?? baseURL
    }

    // MARK: 프로젝트

    func listProjects() async throws -> [ProjectView] {
        struct Response: Decodable { let projects: [ProjectView] }
        return try await send(request(path: "projects"), as: Response.self).projects
    }

    /// PC의 `BRIDGE_WORKSPACE_ROOT` 아래에 새 Git 프로젝트를 생성
    func createProject(displayName: String, agentId: String) async throws -> ProjectView {
        try await sendJSON(
            path: "projects",
            body: [
                "displayName": displayName,
                "agent": agentId,
            ],
            as: ProjectView.self
        )
    }

    func listAgents() async throws -> [AgentView] {
        struct Response: Decodable { let agents: [AgentView] }
        return try await send(request(path: "agents"), as: Response.self).agents
    }

    func conversations(projectId: String) async throws -> [ConversationView] {
        struct Response: Decodable { let conversations: [ConversationView] }
        return try await send(
            request(path: "projects/\(projectId)/conversations"),
            as: Response.self
        ).conversations
    }

    func createConversation(projectId: String) async throws -> ConversationView {
        try await sendJSON(
            path: "projects/\(projectId)/conversations",
            body: ["title": "새 대화"],
            as: ConversationView.self
        )
    }

    func deleteConversation(
        projectId: String, conversationId: String
    ) async throws -> ConversationView {
        try await send(
            request(
                path: "projects/\(projectId)/conversations/\(conversationId)",
                method: "DELETE"
            ),
            as: ConversationView.self
        )
    }

    func conversation(projectId: String, conversationId: String) async throws -> ConversationDetail {
        try await send(
            request(path: "projects/\(projectId)/conversations/\(conversationId)"),
            as: ConversationDetail.self
        )
    }

    func startPreview(projectId: String) async throws -> PreviewView {
        try await send(
            request(path: "projects/\(projectId)/preview", method: "POST"),
            as: PreviewView.self
        )
    }

    // MARK: 작업

    /// 그림 전송
    ///   - snapshot: 실제 WKWebView 렌더와 그 위의 획을 같은 좌표계로 보냄
    ///   - clientTaskId: 재전송해도 작업이 중복 생성되지 않게 함
    ///     같은 그림을 다시 보낼 땐 같은 값을 유지하도록 함
    func createTask(
        projectId: String,
        snapshot: CanvasComposer.Snapshot,
        typedNote: String? = nil,
        clientTaskId: String,
        conversationId: String? = nil,
        agentId: String? = nil,
        latencyOptimized: Bool = true
    ) async throws -> TaskCreated {
        var fields = [
            "projectId": projectId,
            "clientTaskId": clientTaskId,
            "origin": "ipad",
            "mode": latencyOptimized ? "visual-fast" : "visual-full",
        ]
        // 짧은 UI 수정 피드백은 CLI 기본 설정을 그대로 사용하지 않음
        if latencyOptimized { fields["effort"] = "low" }
        if let conversationId { fields["conversationId"] = conversationId }
        if let agentId { fields["agentId"] = agentId }
        if let typedNote, !typedNote.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            fields["typedNote"] = typedNote
        }

        var files = [(name: "canvasImage", payload: snapshot.canvas)]
        if let base = snapshot.base {
            files.append((name: "renderedViewImage", payload: base))
        }

        let (body, contentType) = Self.multipartBody(fields: fields, files: files)
        var req = request(path: "tasks", method: "POST")
        req.setValue(contentType, forHTTPHeaderField: "Content-Type")
        req.httpBody = body
        return try await send(req, as: TaskCreated.self)
    }

    /// 이미지 없이 텍스트만 전송
    func createTask(
        projectId: String,
        typedNote: String,
        clientTaskId: String,
        conversationId: String? = nil,
        agentId: String? = nil,
        latencyOptimized: Bool = true
    ) async throws -> TaskCreated {
        var fields = [
            "projectId": projectId,
            "clientTaskId": clientTaskId,
            "typedNote": typedNote,
            "origin": "ipad",
        ]
        if latencyOptimized { fields["effort"] = "low" }
        if let conversationId { fields["conversationId"] = conversationId }
        if let agentId { fields["agentId"] = agentId }
        let (body, contentType) = Self.multipartBody(fields: fields, files: [])
        var req = request(path: "tasks", method: "POST")
        req.setValue(contentType, forHTTPHeaderField: "Content-Type")
        req.httpBody = body
        return try await send(req, as: TaskCreated.self)
    }

    func task(_ taskId: String) async throws -> TaskView {
        try await send(request(path: "tasks/\(taskId)"), as: TaskView.self)
    }

    func tasks(projectId: String, limit: Int = 30) async throws -> [TaskView] {
        struct Response: Decodable { let tasks: [TaskView] }
        let req = request(path: "tasks", query: [
            URLQueryItem(name: "projectId", value: projectId),
            URLQueryItem(name: "limit", value: String(limit)),
        ])
        return try await send(req, as: Response.self).tasks
    }

    /// 사용자의 모호한 질문에 대하여 LLM이 되물음 -> 추천 선택지 또는 사용자가 직접 입력한 문장으로 답 가능
    func answer(
        _ taskId: String,
        questionId: String,
        optionId: String? = nil,
        freeText: String? = nil
    ) async throws -> TaskCreated {
        var body: [String: Any] = ["questionId": questionId]
        if let optionId { body["selectedOptionId"] = optionId }
        if let freeText,
           !freeText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            body["freeText"] = freeText
            if optionId == nil { body["selectedOptionId"] = "__free_text__" }
        }
        return try await sendJSON(
            path: "tasks/\(taskId)/answer",
            body: body,
            as: TaskCreated.self
        )
    }

    func cancel(_ taskId: String) async throws -> TaskView {
        try await send(request(path: "tasks/\(taskId)/cancel", method: "POST"), as: TaskView.self)
    }

    // MARK: - 내부

    private func request(
        path: String, method: String = "GET", query: [URLQueryItem] = []
    ) -> URLRequest {
        let url = serverRootURL.appendingPathComponent("\(Self.prefix)/\(path)")
        var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        if !query.isEmpty { components?.queryItems = query }

        var req = URLRequest(url: components?.url ?? url)
        req.httpMethod = method
        return req
    }

    private func sendJSON<T: Decodable>(
        path: String, body: [String: Any], as type: T.Type
    ) async throws -> T {
        var req = request(path: path, method: "POST")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        return try await send(req, as: type)
    }

    private func send<T: Decodable>(_ request: URLRequest, as type: T.Type) async throws -> T {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw BridgeError.network(error)
        }

        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            throw BridgeError(message: Self.message(for: status, body: data), statusCode: status)
        }

        do {
            let decoder = JSONDecoder()
            decoder.dateDecodingStrategy = .custom { decoder in
                let container = try decoder.singleValueContainer()
                let value = try container.decode(String.self)
                let fractional = ISO8601DateFormatter()
                fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
                if let date = fractional.date(from: value) { return date }
                let standard = ISO8601DateFormatter()
                standard.formatOptions = [.withInternetDateTime]
                if let date = standard.date(from: value) { return date }
                throw DecodingError.dataCorruptedError(
                    in: container,
                    debugDescription: "ISO 8601 날짜를 해석할 수 없습니다: \(value)"
                )
            }
            return try decoder.decode(T.self, from: data)
        } catch {
            throw BridgeError(message: "서버 응답을 해석할 수 없습니다.", statusCode: status)
        }
    }

    private static func message(for status: Int, body: Data) -> String {
        if let detail = try? JSONDecoder().decode(ErrorDetail.self, from: body).detail,
           !detail.isEmpty {
            return detail
        }
        switch status {
        case 401: return "Tailscale 연결과 로그인 계정을 확인해 주세요."
        case 404: return "요청한 항목을 iMac에서 찾을 수 없습니다."
        case 409: return "지금은 처리할 수 없는 상태입니다."
        case 413: return "보낸 이미지가 너무 큽니다."
        case 415: return "지원하지 않는 이미지 형식입니다."
        case 503: return "iMac에서 아직 준비되지 않은 기능입니다."
        default: return "요청이 실패했습니다(\(status))."
        }
    }

    // MARK: multipart 조립

    private static func multipartBody(
        fields: [String: String],
        files: [(name: String, payload: CanvasComposer.ImagePayload)]
    ) -> (Data, String) {
        let boundary = "Boundary-\(UUID().uuidString)"
        var body = Data()

        for (key, value) in fields {
            body.append("--\(boundary)\r\n")
            body.append("Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n")
            body.append("\(value)\r\n")
        }

        for file in files {
            body.append("--\(boundary)\r\n")
            body.append(
                "Content-Disposition: form-data; name=\"\(file.name)\";"
                + " filename=\"\(file.payload.filename)\"\r\n"
            )
            body.append("Content-Type: \(file.payload.mimeType)\r\n\r\n")
            body.append(file.payload.data)
            body.append("\r\n")
        }

        body.append("--\(boundary)--\r\n")
        return (body, "multipart/form-data; boundary=\(boundary)")
    }
}

private extension Data {
    mutating func append(_ string: String) {
        if let data = string.data(using: .utf8) { append(data) }
    }
}

File: flutter-app/lib/api/bridge_client.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart';

import 'models.dart';

/// 서버가 주는 안내 문구를 그대로 살린다.
class BridgeException implements Exception {
  const BridgeException(this.message, [this.statusCode]);
  final String message;
  final int? statusCode;

  @override
  String toString() => message;
}

/// PC의 Desktop Bridge와 통신한다.
class BridgeClient {
  BridgeClient(String baseUrl) : root = _normalizeRoot(baseUrl);

  /// `/api/v1/...`가 섞여 들어와도 서버 루트로 되돌린다.
  static Uri _normalizeRoot(String input) {
    var uri = Uri.parse(input.trim());
    if (!uri.hasScheme) uri = Uri.parse('http://${input.trim()}');
    final segments = List.of(uri.pathSegments);
    final apiIndex = segments.indexWhere((s) => s.toLowerCase() == 'api');
    if (apiIndex >= 0 &&
        apiIndex + 1 < segments.length &&
        segments[apiIndex + 1].toLowerCase() == 'v1') {
      segments.removeRange(apiIndex, segments.length);
    }
    return uri.replace(pathSegments: segments, query: null, fragment: null);
  }

  final Uri root;
  final http.Client _http = http.Client();

  Uri _uri(String path, [Map<String, String>? query]) => root.replace(
        pathSegments: [
          ...root.pathSegments.where((s) => s.isNotEmpty),
          'api',
          'v1',
          ...path.split('/'),
        ],
        queryParameters: query,
      );

  // ── 프로젝트 ──────────────────────────────────────

  Future<int> health() async {
    final json = await _get('health');
    return json['projects'] as int? ?? 0;
  }

  Future<List<ProjectView>> listProjects() async {
    final json = await _get('projects');
    return [
      for (final p in json['projects'] as List)
        ProjectView.fromJson(p as Map<String, dynamic>),
    ];
  }

  Future<PreviewView> startPreview(String projectId) async {
    final json = await _post('projects/$projectId/preview');
    return PreviewView.fromJson(json);
  }

  // ── 작업 ──────────────────────────────────────────

  /// 그림을 보낸다. canvasPng는 획만 있는 투명 PNG.
  /// renderedJpeg(라이브 렌더 캡처)는 아직 없으면 생략 가능.
  Future<TaskCreated> createTask({
    required String projectId,
    required Uint8List canvasPng,
    required String clientTaskId,
    Uint8List? renderedJpeg,
    String? typedNote,
    String? conversationId,
    String? agentId,
  }) async {
    final req = http.MultipartRequest('POST', _uri('tasks'))
      ..fields['projectId'] = projectId
      ..fields['clientTaskId'] = clientTaskId;
    if (conversationId != null) req.fields['conversationId'] = conversationId;
    if (agentId != null) req.fields['agentId'] = agentId;
    if (typedNote != null && typedNote.trim().isNotEmpty) {
      req.fields['typedNote'] = typedNote;
    }
    req.files.add(http.MultipartFile.fromBytes(
      'canvasImage',
      canvasPng,
      filename: 'canvas.png',
      contentType: MediaType('image', 'png'),
    ));
    if (renderedJpeg != null) {
      req.files.add(http.MultipartFile.fromBytes(
        'renderedViewImage',
        renderedJpeg,
        filename: 'base.jpg',
        contentType: MediaType('image', 'jpeg'),
      ));
    }
    return TaskCreated.fromJson(await _send(req));
  }

  Future<TaskView> task(String taskId) async =>
      TaskView.fromJson(await _get('tasks/$taskId'));

  /// 되물음에 추천 선택지 또는 직접 입력으로 답한다.
  Future<TaskCreated> answer(
    String taskId, {
    required String questionId,
    String? optionId,
    String? freeText,
  }) async {
    final body = <String, dynamic>{'questionId': questionId};
    if (optionId != null) body['selectedOptionId'] = optionId;
    if (freeText != null && freeText.trim().isNotEmpty) {
      body['freeText'] = freeText;
      // 이전 백엔드 호환: 직접 입력임을 나타내는 예약값.
      if (optionId == null) body['selectedOptionId'] = '__free_text__';
    }
    return TaskCreated.fromJson(await _postJson('tasks/$taskId/answer', body));
  }

  Future<TaskView> cancel(String taskId) async =>
      TaskView.fromJson(await _post('tasks/$taskId/cancel'));

  // ── 내부 ──────────────────────────────────────────

  Future<Map<String, dynamic>> _get(String path) async {
    try {
      return _decode(await _http.get(_uri(path)));
    } on SocketException {
      throw const BridgeException('PC에 연결할 수 없습니다. 브리지 서버가 켜져 있는지 확인해 주세요.');
    }
  }

  Future<Map<String, dynamic>> _post(String path) async {
    try {
      return _decode(await _http.post(_uri(path)));
    } on SocketException {
      throw const BridgeException('PC에 연결할 수 없습니다. 브리지 서버가 켜져 있는지 확인해 주세요.');
    }
  }

  Future<Map<String, dynamic>> _postJson(
      String path, Map<String, dynamic> body) async {
    try {
      return _decode(await _http.post(
        _uri(path),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode(body),
      ));
    } on SocketException {
      throw const BridgeException('PC에 연결할 수 없습니다. 브리지 서버가 켜져 있는지 확인해 주세요.');
    }
  }

  Future<Map<String, dynamic>> _send(http.MultipartRequest req) async {
    try {
      final streamed = await _http.send(req);
      return _decode(await http.Response.fromStream(streamed));
    } on SocketException {
      throw const BridgeException('PC에 연결할 수 없습니다. 브리지 서버가 켜져 있는지 확인해 주세요.');
    }
  }

  Map<String, dynamic> _decode(http.Response res) {
    if (res.statusCode < 200 || res.statusCode >= 300) {
      throw BridgeException(_message(res), res.statusCode);
    }
    try {
      return jsonDecode(utf8.decode(res.bodyBytes)) as Map<String, dynamic>;
    } catch (_) {
      throw BridgeException('서버 응답을 해석할 수 없습니다.', res.statusCode);
    }
  }

  /// 서버가 내려준 detail을 우선 쓰고, 없을 때만 상태 코드로 문구를 만든다.
  String _message(http.Response res) {
    try {
      final detail = (jsonDecode(utf8.decode(res.bodyBytes))
          as Map<String, dynamic>)['detail'] as String?;
      if (detail != null && detail.isNotEmpty) return detail;
    } catch (_) {}
    return switch (res.statusCode) {
      401 => 'Tailscale 연결과 로그인 계정을 확인해 주세요.',
      404 => '요청한 항목을 PC에서 찾을 수 없습니다.',
      409 => '지금은 처리할 수 없는 상태입니다.',
      413 => '보낸 이미지가 너무 큽니다.',
      415 => '지원하지 않는 이미지 형식입니다.',
      503 => 'PC에서 아직 준비되지 않은 기능입니다.',
      _ => '요청이 실패했습니다(${res.statusCode}).',
    };
  }
}
